staking/slash/double-sign.go

```diff
@@ -1,52 +1,341 @@
+	)
+	errSlashDebtNotFullyAccountedFor = errors.New(
+		"slash debt was not fully accounted for, still non-zero",
+	)
+	errShardIDNotKnown              = errors.New("nil subcommittee for shardID")
+	errValidatorNotFoundDuringSlash = errors.New("validator not found")
+	zero                            = numeric.ZeroDec()
+	oneDoubleSignerRate             = numeric.MustNewDecFromStr("0.02")
+	oneHundredPercent               = numeric.NewDec(1)
+	fiftyPercent                    = numeric.MustNewDecFromStr("0.5")
+)
+
+// applySlashRate returns (amountPostSlash, amountOfReduction, amountOfReduction / 2)
+func applySlashRate(amount *big.Int, rate numeric.Dec) *big.Int {
+	return numeric.NewDecFromBigInt(
+		amount,
+	).Mul(rate).TruncateInt()
+}
+
+func delegatorSlashApply(
+	snapshot, current *staking.ValidatorWrapper,
+	rate numeric.Dec,
+	state *state.DB,
+	reporter common.Address,
+	doubleSignEpoch *big.Int,
+	slashTrack *Application,
+) error {
+	for _, delegationSnapshot := range snapshot.Delegations {
+		slashDebt := applySlashRate(delegationSnapshot.Amount, rate)
+		halfOfSlashDebt := new(big.Int).Div(slashDebt, common.Big2)
+		snapshotAddr := delegationSnapshot.DelegatorAddress
+		for _, delegationNow := range current.Delegations {
+			if nowAmt := delegationNow.Amount; delegationNow.DelegatorAddress == snapshotAddr {
+				utils.Logger().Info().
+					RawJSON("delegation-snapshot", []byte(delegationSnapshot.String())).
+					RawJSON("delegation-current", []byte(delegationNow.String())).
+					Uint64("slash-debt", slashDebt.Uint64()).
+					Str("rate", rate.String()).
+					Msg("attempt to apply slashing based on snapshot amount to current state")
+
+				state.AddBalance(reporter, halfOfSlashDebt)
+				// NOTE only need to pay snitch here
+				slashTrack.TotalSnitchReward.Add(
+					slashTrack.TotalSnitchReward, halfOfSlashDebt,
+				)
+				// Current delegation has some money and slashdebt is still not paid off
+				// so contribute as much as can with current delegation amount
+				if nowAmt.Cmp(common.Big0) == 1 && slashDebt.Cmp(common.Big0) == 1 {
+					// 0.50_amount > 0.06_debt => slash == 0.0, nowAmt == 0.44
+					if nowAmt.Cmp(slashDebt) >= 0 {
+						nowAmt.Sub(nowAmt, slashDebt)
+						payDebt(
+							snapshot, current, slashDebt, slashDebt, slashTrack,
+						)
+					} else {
+						// 0.50_amount < 2.4_debt =>, slash == 1.9, nowAmt == 0.0
+						payDebt(
+							snapshot, current, slashDebt, nowAmt, slashTrack,
+						)
+						nowAmt.Sub(nowAmt, nowAmt)
+					}
+				}
+
+				// NOTE Assume did as much as could above, now check the undelegations
+				for _, undelegate := range delegationNow.Undelegations {
+					// the epoch matters, only those undelegation
+					// such that epoch>= doubleSignEpoch should be slashable
+					if undelegate.Epoch.Cmp(doubleSignEpoch) >= 0 {
+						if slashDebt.Cmp(common.Big0) <= 0 {
+							// paid off the slash debt
+							break
+						}
+
+						if undelegate.Amount.Cmp(common.Big0) == 1 &&
```

Comment:
This logic is still wrong. There should be two cases. One is debt >= undelegation.amount, then take out the whole amount and remove it, go to next undelegation. Otherwise, take the "debt" amount on the undelegation and break.